fastavro/_write_py.py

```diff
@@ -296,89 +312,174 @@ def write_map(fo, datum, schema):
-        )
+    )
 
-    if record_type == 'union':
-        if isinstance(datum, tuple):
-            (name, datum) = datum
-            for candidate in schema:
-                if extract_record_type(candidate) == 'record':
-                    if name == candidate["name"]:
-                        return validate(datum, candidate)
-            else:
-                return False
-        return any(validate(datum, s) for s in schema)
 
-    # dict-y types from here on.
-    if record_type == 'enum':
-        return datum in schema['symbols']
+def validate_enum(datum, schema, **kwargs):
+    return datum in schema['symbols']
 
-    if record_type == 'array':
-        return (
+
+def validate_array(datum, schema, raise_errors=False):
+    return (
             isinstance(datum, Iterable) and
             not is_str(datum) and
-            all(validate(d, schema['items']) for d in datum)
-        )
+            all(validate(datum=d, schema=schema['items'],
+                         field=schema.get('name'),
+                         raise_errors=raise_errors) for d in datum)
+    )
+
 
-    if record_type == 'map':
-        return (
+def validate_map(datum, schema, raise_errors=False):
+    return (
             isinstance(datum, Mapping) and
             all(is_str(k) for k in iterkeys(datum)) and
-            all(validate(v, schema['values']) for v in itervalues(datum))
-        )
+            all(validate(datum=v, schema=schema['values'],
+                         field=schema.get('name'),
+                         raise_errors=raise_errors) for v in itervalues(datum))
+    )
+
 
-    if record_type in ('record', 'error', 'request',):
-        return (
+def validate_record(datum, schema, raise_errors=False):
+    return (
             isinstance(datum, Mapping) and
             all(
-                validate(datum.get(f['name'], f.get('default')), f['type'])
+                validate(datum=datum.get(f['name'], f.get('default')),
+                         schema=f['type'],
+                         field=schema.get('name'),
+                         raise_errors=raise_errors)
                 for f in schema['fields']
             )
-        )
+    )
 
-    if record_type in SCHEMA_DEFS:
-        return validate(datum, SCHEMA_DEFS[record_type])
 
-    raise ValueError('unkown record type - %s' % record_type)
+def validate_union(datum, schema, raise_errors=False):
+    if isinstance(datum, tuple):
+        (name, datum) = datum
+        for candidate in schema:
+            if extract_record_type(candidate) == 'record':
+                if name == candidate["name"]:
+                    return validate(datum, schema=candidate,
+                                    field=None,
+                                    raise_errors=raise_errors)
+        else:
+            return False
+
+    errors = []
+    for s in schema:
+        try:
+            ret = validate(datum, schema=s,
+                           field=None,
```

Comment:
I think this causes the eventual exception (if raised) to have a field value of `None`. Is this intended?